tests/unit/test_charm.py

```diff
@@ -926,3 +926,186 @@ def test_given_service_is_running_when_on_remove_then_service_is_stopped(
+        )
+        (
+            app_name,
+            host_ip,
+            _,
+            rel_id,
+            _,
+        ) = self.setup_vault_kv_relation(nb_units=3)
+
+        configure_approle.return_value = "12345678"
+        generate_role_secret_id.return_value = "11111111"
+
+        mount_suffix = "dummy"
+        self.harness.update_relation_data(rel_id, app_name, {"mount_suffix": mount_suffix})
+
+        relation_data = self.harness.get_relation_data(rel_id, self.harness.charm.app.name)
+        assert relation_data["vault_url"] == f"https://{host_ip}:{self.harness.charm.VAULT_PORT}"
+        assert relation_data["mount"] == "charm-" + app_name + "-" + mount_suffix
+        for secret_id in json.loads(relation_data["credentials"]).values():
+            secret = self.harness.model.get_secret(id=secret_id)
+            secret_content = secret.get_content()
+            assert configure_approle.return_value == secret_content["role-id"]
+            assert generate_role_secret_id.return_value == secret_content["role-secret-id"]
+
+    @patch("charms.vault_k8s.v0.vault_kv.VaultKvProvides.set_unit_credentials")
+    @patch("charms.vault_k8s.v0.vault_kv.VaultKvProvides.set_ca_certificate")
+    @patch("charms.vault_k8s.v0.vault_kv.VaultKvProvides.set_vault_url")
+    @patch("charms.vault_k8s.v0.vault_kv.VaultKvProvides.set_mount")
+    @patch("vault.Vault.generate_role_secret_id")
+    @patch("vault.Vault.configure_approle")
+    @patch("vault.Vault.configure_kv_policy")
+    @patch("vault.Vault.configure_kv_mount")
+    @patch("vault.Vault.is_api_available")
+    def test_given_unit_is_not_leader_when_secret_kv_is_complete_then_no_data_is_updated(
+        self,
+        _,
+        __,
+        ___,
+        configure_approle,
+        generate_role_secret_id,
+        set_mount,
+        set_vault_url,
+        set_ca_certificate,
+        set_unit_credentials,
+    ):
+        (
+            app_name,
+            _,
+            _,
+            rel_id,
+            _,
+        ) = self.setup_vault_kv_relation(nb_units=3)
+        self.harness.set_leader(False)
+
+        configure_approle.return_value = "12345678"
+        generate_role_secret_id.return_value = "11111111"
+
+        mount_suffix = "dummy"
+        self.harness.update_relation_data(rel_id, app_name, {"mount_suffix": mount_suffix})
+
+        set_mount.assert_not_called()
+        set_vault_url.assert_not_called()
+        set_ca_certificate.assert_not_called()
+        set_unit_credentials.assert_not_called()
+
+    @patch("vault.Vault.read_role_secret")
+    @patch("vault.Vault.generate_role_secret_id")
+    @patch("vault.Vault.configure_approle")
+    @patch("vault.Vault.configure_kv_policy")
+    @patch("vault.Vault.configure_kv_mount")
+    @patch("vault.Vault.is_api_available")
+    @patch("vault.Vault.enable_approle_auth")
+    def test_given_unit_is_leader_when_related_unit_egress_is_updated_then_secret_content_is_updated(  # noqa: E501
```

Comment:
This test does too many things, let's make sure that we only have asserts for what we are actually validating. If we need to validate multiple behavior it's a sign that we need multiple tests

Reply:
Reduced the content of this test in last commit.